src/m365/spo/commands/file/file-list.ts

```diff
@@ -96,20 +130,80 @@ class SpoFileListCommand extends SpoCommand {
     };
 
     return request
-      .get<FileFolderCollection>(requestOptions)
-      .then((filesAndFoldersResult: FileFolderCollection) => {
-        filesAndFoldersResult.Files.forEach((file: FileProperties) => files.value.push(file));
+      .get<{ value: FileProperties[] }>(requestOptions)
+      .then(async (filesAndFoldersResult: { value: FileProperties[] }) => {
+        filesAndFoldersResult.value.forEach((file: FileProperties) => files.value.push(file));
         // If the request is --recursive, call this method for other folders.
         if (args.options.recursive &&
-          filesAndFoldersResult.Folders !== undefined &&
-          filesAndFoldersResult.Folders.length !== 0) {
-          return Promise.all(filesAndFoldersResult.Folders.map((folder: { ServerRelativeUrl: string; }) => this.getFiles(folder.ServerRelativeUrl, args, files)));
+          subFolders.length !== 0) {
+          return Promise.all(subFolders.map(async (folder: string) => {
+            const folderProperties = await this.getItemCount(folder, args);
+            this.getFiles(folder, args, folderProperties.folders, folderProperties.items, 0, files);
+          }));
+        }
+        else if (items > SpoFileListCommand.tresholdLimit && (items - index > SpoFileListCommand.tresholdLimit)) {
+          return await this.getFiles(folderUrl, args, [], items, index + SpoFileListCommand.tresholdLimit, files);
         }
         else {
           return;
         }
       }).then(() => files);
   }
+
+  private async getItemCount(folderUrl: string, args: CommandArgs): Promise<{ items: number, folders: any[] }> {
+    let expandProperties = 'Properties';
+    if (args.options.recursive) {
```

Comment:
I was doing this here to combine the count and folders call. The idea behind this is that we have to do 2 api calls per folder level (not taking into account doing more api calls when the threshold is exceeded).
- one for the count and the subfolders (which you would do on every folder level since a folder can also have 5000 items or more)
- one for the actual files.

I'm pretty sure we see it the same way, but yours is a different approach and as I understand based on your explanation in below review comment, this would result in 3 api calls per folder level:
- one for the count
- one for the subfolders
- one for the files

You see what my point behind this way of working is?

Reply:
Hi @nicodecleyre, if we use the count endpoint we have to do that on every subfolder as well, otherwise it's not really useful. 

But thinking about it: The count http-request might save us HTTP requests to the server in certain situations, because you know in time you can stop querying (for files as well as the folders endpoint). But on the other hand, in many situations it may just be an extra call. Having slept on it a night: Let's disregard it and go for least complexity.

Just call the files and folders endpoints on all folders. (and make sure you follow it to the end, using $skip and $top. The check is: if a query returns as many items as were requested (result.count equals $top) , run another query to get the next page.

Reply:
And thank YOU for all your time and dedication to the CLI project!